### ISSUE TITLE
unit-no-unknown

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#1004]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Throw error if found unknown unit. Add options `ignore` to define own unit.

> Which rule, if any, is this issue related to?

`unit-no-unknown`

> What CSS is needed to reproduce this issue?

```css
.foo {
  width: 100qw;
}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
  "rules": {
    "unit-no-unknown": true
  }
}
```

> Which version of stylelint are you using?

`5.2.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

no matter

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

no matter

> What did you expect to happen?

Throw error if found unknown units

> What actually happened (e.g. what warnings or errors you are getting)?

Nothing

**Comment by jeddy3** [comment]
Seems like a great companion to `property-no-unknown`.

Just checking that we are all comfortable the `*-no-unknown` rules co-existing with the `*-whitelist` rules e.g. `unit-no-unknown` and `unit-whitelist`. It's unlikely the rules would be used together as a user-defined whitelist would serve the same purpose as the `*-no-unknown` rule.

**Comment by davidtheclark** [comment]
Is `property-no-unknown` the same as `property-whitelist` with a preset list?

If so, maybe we convert the `*-no-unknown` proposals into *preset lists for the `*-whitelist` rules. Maybe `property-whitelist` accepts a keyword primary option `"standard"` that loads a preset list of standard properties? Same would go for other `*-no-unknown` proposals, like this one. What do you think?

**Comment by jeddy3** [comment]
> Is property-no-unknown the same as property-whitelist with a preset list?

Yep, that's what I realised.

> What do you think?

I think this might be the way forward. Would it be odd to still allow a `ignoreProperties` secondary option with `property-whitelist`? e.g.

```js
"property-whitelist": ["standard", { ignoreProperties: ["lost-grid"] }]
```

As well as:

```js
"property-whitelist": ["flex", "align-items", ...]
```

**Comment by evilebottnawi** [comment]
@jeddy3 Presets this is a very good solution, then perhaps we will not have to create a whole set of rules `*-no-unknown` for properties, values, and units and etc (example media features, media types).

**Comment by jeddy3** [comment]
I just had a thought. 

Would moving including `standard` properties to the secondary option be clearer? e.g.

```js
"property-whitelist": [["compose"], { include: ["standard-properties"] }]
```

Or is it a bit odd if having an empty array if you only want standard properties e.g.

```js
"property-whitelist": [[], { include: ["standard-properties"] }]
```

It sounds like we're all in favour of repurposing these `*-whitelist` rules... it's just deciding which options configuration is neatest. I think I'm in favour if this configuration, rather than the `ignoreProperties: ["lost-grid"] }` as it feels odd having a secondary option that can be redundant.

What do you guys think? Once we're decided we'll get this issue and https://github.com/stylelint/stylelint/issues/950 marked as "help wanted".

**Comment by davidtheclark** [comment]
Do we currently have `ignoreProperties`? I don't see that documented: http://stylelint.io/user-guide/rules/property-whitelist/

**Comment by jeddy3** [comment]
Nope, not yet. It [comes from](https://github.com/stylelint/stylelint/issues/950#issuecomment-200384045) the `property-no-unknown` proposal i.e. it's a way of ignoring non-standard properties in that proposal. 

I think:

```js
"property-whitelist": [["compose"], { include: ["standard-properties"] }]
```

might be a better way of achieving the same though.

**Comment by davidtheclark** [comment]
Ok, I was confused.

There's just one problem I have in mind. What if I want to include `standard-properties` but slightly modified to *remove* some of the properties in the list?

**Comment by jeddy3** [comment]
Good point. I guess you could use the corresponding `*-blacklist` rule? e.g

```js
"property-whitelist": [["compose"], { include: ["standard-properties"] }],
"property-blacklist": ["float"]
```

But is it starting to feel messy now i.e. I'd say it's more intuitive to have a whitelist *or* blacklist, rather than being able to use both?

Perhaps separate `*-no-unknown` rules are the way to go e.g. if you're wanting to go down the blacklist route:

```js
"property-no-unknown": [true, { includeProperties: ["compose"] }],
"property-blacklist": ["float"]
```

And if you're wanting to be restrictive (i.e. only allow a small subset of properties, for whatever reason) from the start then you'd probably just reach for the `*-whitelist` rules e.g.

```js
"property-whitelist: ["display", "margin", "padding"]
```

It feels like the former is the more likely use case.

**Comment by davidtheclark** [comment]
I was thinking the same thing: now that it comes down to the nitty-gritty of choosing options, I do think that separate rules might be better. Full circle!

**Comment by jeddy3** [comment]
> Full circle!

Ha, yep! I'm glad we thrashed it out though as I feel we're more confident in our understanding of the use case of the `*-whitelist`, `*-blacklist` and `*-no-unknown` rules and how they all work together :)

> comes down to the nitty-gritty of choosing options

I think `ignoreProperties: []` on just the `*-no-unknown` rules covers all of the bases. The `*-whitelist` and `*-blacklist` rules remain untouched. `ignoreProperties` [is inline with our options conventions](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#determine-options).

If we're in agreement, I'll reapply the "rule" and "help" wanted labels.

**Comment by evilebottnawi** [comment]
@jeddy3 separate rule? i can made this rule on this week

**Comment by davidtheclark** [comment]
Yeah we are in agreement.

**Comment by jeddy3** [comment]
@evilebottnawi Yep, separate rule.

- Name: `property-no-unknown`
- Primary option: `true`.
- Secondary options: `ignoreProperties: []` (user defined array of unprefixed property names)
- Message: "Unexpected unknown property".

> i can made this rule on this week

Fantastic! :)
